Add IPC for SkData

donny-dont opened the pull request:
#### 7f78e4799e9778e292febfbab858754745c580ac
<pre>
Add IPC for SkData
<a href="https://bugs.webkit.org/show_bug.cgi?id=273890">https://bugs.webkit.org/show_bug.cgi?id=273890</a>

Reviewed by Alex Christensen.

Add a wrapper of `SkData` for serialization over IPC.

* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/skia/CoreIPCSkData.h: Added.
(WebKit::CoreIPCSkData::CoreIPCSkData):
(WebKit::CoreIPCSkData::skData const):
(WebKit::CoreIPCSkData::dataReference const):
* Source/WebKit/Shared/skia/CoreIPCSkData.serialization.in: Added.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
(IPC::ArgumentCoder&lt;sk_sp&lt;SkData&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;sk_sp&lt;SkData&gt;&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/278536@main">https://commits.webkit.org/278536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/170e49fbd0b874d6b74d74fe4745629bd33b1f62

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50875 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30174 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54134 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1566 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36448 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1220 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52974 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27812 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43825 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/22576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25150 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9316 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47135 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1151 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55728 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25978 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/1030 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48848 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27235 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43898 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11145 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26966 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/skia/CoreIPCSkData.h:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    CoreIPCSkData(std::span<const uint8_t> data)
+        : m_skData(SkData::MakeWithoutCopy(data.data(), data.size()))
```
I'm not sure the "WithoutCopy" is safe in all cases.

donny-dont replied:
Fixed

achristensen07 commented on Source/WebKit/Shared/WebCoreArgumentCoders.h:
```diff
@@ -76,6 +77,12 @@ template<> struct ArgumentCoder<sk_sp<SkColorSpace>> {
     static std::optional<sk_sp<SkColorSpace>> decode(Decoder&);
 };
+
+template<> struct ArgumentCoder<sk_sp<SkData>> {
```
This file should be decreasing until it is removed.  *.serialization.in has "wrapped by" to accomplish this.  It may need a little adapting.

donny-dont replied:
I am with you on this goal but that will require some surgery. Is it ok to land this and open a bug to make `wrapped by` work on non-cocoa situations?

donny-dont replied:
https://bugs.webkit.org/show_bug.cgi?id=273904

donny-dont committed:
Add IPC for SkData
https://bugs.webkit.org/show_bug.cgi?id=273890

Reviewed by Alex Christensen.

Add a wrapper of `SkData` for serialization over IPC.

* Source/WebKit/Platform/Skia.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/skia/CoreIPCSkData.h: Added.
(WebKit::CoreIPCSkData::CoreIPCSkData):
(WebKit::CoreIPCSkData::skData const):
(WebKit::CoreIPCSkData::dataReference const):
* Source/WebKit/Shared/skia/CoreIPCSkData.serialization.in: Added.
* Source/WebKit/Shared/skia/WebCoreArgumentCodersSkia.cpp:
(IPC::ArgumentCoder<sk_sp<SkData>>::encode):
(IPC::ArgumentCoder<sk_sp<SkData>>::decode):

Canonical link: https://commits.webkit.org/278536@main
